### PR TITLE
Sota ccl

### PR DESCRIPTION
**Pull request by prittt**
This PR introduces optimization for `connectedComponents()`  and `connectedComponentsWithStats()` functions by adding a new state-of-the-art CCL algorithm (i.e. Spaghetti Labeling) as described in [Spaghetti Labeling: Directed Acyclic Graphs for Block-Based Connected Components Labeling, IEEE Transactions on Image Processing](https://ieeexplore.ieee.org/document/8874953). Performance comparisons can be found in the paper and in the [YACCLAB repository](https://github.com/prittt/YACCLAB). Tests have been updated accordingly and additional test cases for statistics calculation have been added.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by prittt** [comment]
Dear @alalek, could you please check Jenkins build and re-run this job? It fails with a git related issue:

```
stderr: fatal: unable to access 'https://github.com/opencv/opencv.git/': Operation timed out after 300041 milliseconds with 0 out of 0 bytes received
```

Thank you in advance.

**Comment by alalek** [comment]
> Jenkins

/cc @asmorkalov 

---

> Operation timed out

@prittt You can ignore this for now because it is not related to this patch.

**Comment by asmorkalov** [comment]
@prittt Sorry for inconvenience, but Jenkins setup suffers from connectivity issues. We are working on the problem with local IT and hopefully it'll be fixed in mean time. Please ignore Jenkins CI status for now, it's optional.

File: modules/imgproc/src/ccl_bolelli_forest.inc.hpp [hunk 1]
```diff
@@ -0,0 +1,1783 @@
...
+//
+//  IMPORTANT: READ BEFORE DOWNLOADING, COPYING, INSTALLING OR USING.
+//
+//  By downloading, copying, installing or using the software you agree to this license.
+//  If you do not agree to this license, do not download, install,
+//  copy or use the software.
+//
+//
+//                        Intel License Agreement
+//                For Open Source Computer Vision Library
```
Comment (alalek) [inline modules/imgproc/src/ccl_bolelli_forest.inc.hpp:11]: Please use short license header: https://github.com/opencv/opencv/wiki/Coding_Style_Guide#file-structure

```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
//
// 2021 Federico Bolelli <federico.bolelli@unimore.it>
// 2021 Stefano Allegretti <stefano.allegretti@unimore.it>
// 2021 Costantino Grana <costantino.grana@unimore.it>
```

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: modules/imgproc/src/ccl_bolelli_forest.inc.hpp (hunk 1 above)
Comment (prittt) [reply modules/imgproc/src/ccl_bolelli_forest.inc.hpp:11]: @alalek, sorry I've missed it. Do you want me to replace also the [connectedcomponents.cpp](https://github.com/prittt/opencv/blob/sota-ccl/modules/imgproc/src/connectedcomponents.cpp) license header or just the ones in the hpp files?

**Comment by prittt** [comment]
> 
> 
> @prittt Sorry for inconvenience, but Jenkins setup suffers from connectivity issues. We are working on the problem with local IT and hopefully it'll be fixed in mean time. Please ignore Jenkins CI status for now, it's optional.

@asmorkalov no problem. Thank you for the fast answer.

File: modules/imgproc/src/ccl_bolelli_forest.inc.hpp (hunk 1 above)
Comment (alalek) [reply modules/imgproc/src/ccl_bolelli_forest.inc.hpp:11]: Lets update in new files only for now.

**Review by alalek — APPROVED** [review]
Great! Thank you for contribution :+1:

**Pull request merged**